source/annotations/helpers.go

```
@@ -0,0 +1,102 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package annotations
+
+import (
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+func getAliasFromAnnotations(annotations map[string]string) bool {
+	aliasAnnotation, exists := annotations[AliasAnnotationKey]
+	return exists && aliasAnnotation == "true"
+}
+
+// TTLFromAnnotations TODO: copied from source.go. Refactor to avoid duplication.
+// TTLFromAnnotations extracts the TTL from the annotations of the given resource.
+func TTLFromAnnotations(annotations map[string]string, resource string) endpoint.TTL {
+	ttlNotConfigured := endpoint.TTL(0)
+	ttlAnnotation, exists := annotations[TtlAnnotationKey]
+	if !exists {
+		return ttlNotConfigured
+	}
+	ttlValue, err := parseTTL(ttlAnnotation)
+	if err != nil {
+		log.Warnf("%s: \"%v\" is not a valid TTL value: %v", resource, ttlAnnotation, err)
```

IMO it's clearly an error. Warning should have only use of communication deprecations, in my opinion.
Maybe we should decide how we use warnings, because I think during the last couple of months there's a little abuse of warnings.

---

Link to original code https://github.com/kubernetes-sigs/external-dns/blob/277e8ca2ec7e10f749083bfd9842d61cc4fd8f32/source/source.go#L91-L106

And pull commit where change was introduced https://github.com/kubernetes-sigs/external-dns/commit/17e9637f1182f070488c7f6d558b927076e0c413#diff-cf68f602fa7c20e5341f3b83054df68ade1586a144b1eae5347e0ac47096d3aa

The Warning makes absolute sense.
- When parseTTL failes, there is a Warning that user should change TTL configuration
- Code uses default values, external-dns works.
- User could fix annotation or configuration when have time, and Warning will disapear

If it's an error;
- the code most likely should throw an error instead of setting up a default
- If there is a mistake in configuration, the external-dns is going to crash or behave differently

I need to capture all cases and provide an update for annotations proposal https://github.com/kubernetes-sigs/external-dns/pull/5080

It was like that and I would rather keep it as is. In my opinion, I better capture this in annotation graduation proposal. Kind of all annotations should have validations, and validation should either be Error or Warning as example.

---

ok